scheduler/reconciler/reconcile_node.go

```diff
@@ -68,21 +75,54 @@ func (nr *NodeReconciler) Compute(
 	return result
 }
 
-// diffSystemAllocsForNode is used to do a set difference between the target allocations
-// and the existing allocations for a particular node. This returns 8 sets of results,
-// the list of named task groups that need to be placed (no existing allocation), the
-// allocations that need to be updated (job definition is newer), allocs that
-// need to be migrated (node is draining), the allocs that need to be evicted
-// (no longer required), those that should be ignored, those that are lost
-// that need to be replaced (running on a lost node), those that are running on
-// a disconnected node but may resume, and those that may still be running on
-// a node that has resumed reconnected.
+// computeCanaryNodes is a helper function that, given required task groups and
+// eligible nodes, outputs a map of arrays of nodes, indexed by task group name,
+// on which canaries should be placed.
+func computeCanaryNodes(required map[string]*structs.TaskGroup,
+	eligibleNodes map[string]*structs.Node) map[string][]*structs.Node {
+
+	canaryNodes := map[string][]*structs.Node{}
+	eligibleNodesList := slices.Collect(maps.Values(eligibleNodes))
+	for _, tg := range required {
+		if tg.Update.IsEmpty() || tg.Update.Canary == 0 {
+			continue
+		}
+
+		for i, n := range eligibleNodesList {
+			if i > (tg.Update.Canary*len(eligibleNodes))/100 {
```

Comment:
The rounding-up on this is a little odd. Ex. if you have 5 eligible nodes having canary=39 returns 2 nodes but canary=40 returns 3 (60%, even though 40% is a round number out of 5)

Reply:
[f40f050](https://github.com/hashicorp/nomad/pull/26499/commits/f40f0502317e24406c29497320b3bd88bf66020d)

note the commit message :(